### ISSUE TITLE
GITHUB_BEFORE_SHA attempts to reference incorrect commit on merges

### ISSUE THREAD

**Issue by lae** [super-linter/super-linter#5941]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest Super-linter version available?

- [X] I am using the latest Super-linter version.
- [X] I can reproduce the issue running Super-linter using complete version identifier (example: vX.Y.Z), and not just with a partial one (example: vX)

- [X] I am using the super-linter/super-linter action or container image, and not the deprecated github/super-linter action or container image.

### Are you resonably sure that it's a Super-linter issue, and not an issue related to a tool that Super-linter runs?

- [X] I think that this is a Super-linter issue.

### Current Behavior

super-linter attempts to set `GITHUB_BEFORE_SHA` to an older commit than expected when merge commits exist in history, i.e. when PRs are merged into main. For new repos, this causes failures because of an attempt to reference a commit that doesn't exist.

### Expected Behavior

It should set `GITHUB_BEFORE_SHA` to the right commit even in the presence of merges.

### Super-Linter version

```markdown
v6.7.0
```

### Super-linter configuration

```yaml
---
name: Lint Codebase

on:
  pull_request:
  push:
    branches:
      - main

permissions:
  contents: read
  packages: read
  statuses: write

jobs:
  lint:
    name: Lint Codebase
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        id: checkout
        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332  # v4.1.7
        with:
          fetch-depth: 0

      - name: Lint Codebase
        id: super-linter
        uses: super-linter/super-linter/slim@3fe03abab2eafb293ace16d4a3b07aeabcb3f1a0  # v6.7.0
        env:
          DEFAULT_BRANCH: main
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          VALIDATE_ALL_CODEBASE: true
```

### Relevant log output

```shell
standard: 17.1.0
  2024-07-30 20:08:50 [INFO]   --------------------------------------------
  2024-07-30 20:08:50 [INFO]   Gathering GitHub information...
  2024-07-30 20:08:50 [INFO]   Successfully validated GITHUB_WORKSPACE: /github/workspace
  2024-07-30 20:08:50 [INFO]   Successfully found GITHUB_EVENT_PATH: /github/workflow/event.json]
  2024-07-30 20:08:50 [INFO]   Successfully found GITHUB_SHA: 9db5fcdf85d26f5dfd0afbfd4af02c1ca03f99d6
  2024-07-30 20:08:50 [INFO]   Successfully found GITHUB_PUSH_COMMIT_COUNT: 5
  fatal: ambiguous argument 'HEAD~5': unknown revision or path not in the working tree.
  Use '--' to separate paths from revisions, like this:
  'git <command> [<revision>...] -- [<file>...]'
  2024-07-30 20:08:50 [FATAL]   Failed to initialize GITHUB_BEFORE_SHA for a push event. Output: HEAD~5
  /action/lib/linter.sh: line 720: SUPER_LINTER_PRIVATE_OUTPUT_DIRECTORY_PATH: unbound variable
```

### Steps To Reproduce

1. create a PR in repo with any number of commits ([example PR](https://github.com/amplify-security/runner-action/pull/1/commits))
2. super-linter workflow runs on PR and succeeds ([example run](https://github.com/amplify-security/runner-action/actions/runs/10168880670/job/28124525598))
3. merge PR
4. super-linter workflow runs on push and fails (or in most other cases "succeeds" while referencing the wrong commit) ([example run](https://github.com/amplify-security/runner-action/actions/runs/10169011658/job/28124961385#step:4:232))

### Anything else?

The action attempts to reference `HEAD~${GITHUB_PUSH_COMMIT_COUNT}`

https://github.com/super-linter/super-linter/blob/3fe03abab2eafb293ace16d4a3b07aeabcb3f1a0/lib/linter.sh#L444-L446

but `git` appears to treat merges as one, and the correct reference here for a single merge seems to be `HEAD~1` instead.

I think this is the same issue as reported in this https://github.com/super-linter/super-linter/issues/5453#issuecomment-2092661221 but it doesn't seem to be resolved even though it may be silently succeeding for them and others at this point (given their repos probably have enough commits where the `rev-parse` step no longer fails)?

**Comment by ferrarimarco** [comment]
Hi @lae !

Can you please reproduce this locally against a new git repository? So we can first create a test for this.

Thanks!

**Comment by lae** [comment]
Separate repo created here:

https://github.com/amplify-security/sl-repro-5941

I can't reproduce this locally, `GITHUB_BEFORE_SHA` doesn't exist anywhere in the output after running `docker run -e DEFAULT_BRANCH=main -e LOG_LEVEL=DEBUG -e RUN_LOCAL=true -v ./:/tmp/lint --rm ghcr.io/super-linter/super-linter:latest | tee -a run.log`...which I assume is to be expected since this appears to be a Github Actions specific code path?

**Comment by ferrarimarco** [comment]
Thanks!

Your assumption is correct, although we can reach that code path with tests.

I'll have a look at the repo :)

**Comment by ferrarimarco** [comment]
I had a quick look at the reproduction case, that's very helpful. Can you keep it online for a few days more?
I'll have to add a test for repositories that use merge commits.

**Comment by ferrarimarco** [comment]
@lae Can you please re-run the failing workflow in the reproduction example repository with debugging logging enabled? Thanks a lot!

**Comment by lae** [comment]
@ferrarimarco if you still need it, done: https://github.com/amplify-security/sl-repro-5941/actions/runs/10170092001/job/28241370287

**Comment by ferrarimarco** [comment]
Thanks! This is really helpful so I can ensure that the GitHub event data that we pass to the test matches with the one that triggered the workflow.

**Comment by ferrarimarco** [comment]
PS: thanks to your report, we also fixed another bug! #5949

**Comment by ferrarimarco** [comment]
I was able to grab the event JSON from your log, thanks! #5950 should take care of fixing this issue.

If you want, you can go ahead and delete the temp repository where you created the reproduction case. Thanks a lot!
